src/main/java/app/attestation/server/AttestationServer.java

```diff
@@ -1147,6 +1143,104 @@ public void handlePost(final HttpExchange exchange) throws IOException, SQLiteEx
+                    throw new NumberFormatException("Invalid pagination directive provided");
+                }
+                if (!verifyDeviceFingerprint(account.userId, fingerprint)) {
+                    throw new NullPointerException("Invalid device fingerprint provided");
+                }
+            } catch (final ClassCastException | JsonException | NullPointerException | NumberFormatException e) {
+                e.printStackTrace();
+                exchange.sendResponseHeaders(400, -1);
+                return;
+            }
+            writeAttestationHistoryJson(exchange, fingerprint, offsetId, directive);
+        }
+    }
+
+    private static boolean verifyDeviceFingerprint(final long userId, final String deviceFingerprint)
+            throws IOException, SQLiteException {
+        boolean isFingerprintAuthenticated = false;
+        final SQLiteConnection conn = new SQLiteConnection(AttestationProtocol.ATTESTATION_DATABASE);
+        try {
+            final SQLiteStatement select = conn.prepare("SELECT fingerprint FROM Devices WHERE userId is ?");
+            select.bind(1, userId);
+
+            while (select.step()) {
+                String ourFingerprint = BaseEncoding.base16().encode(select.columnBlob(0));
+                if (ourFingerprint.equals(deviceFingerprint)) {
+                    isFingerprintAuthenticated = true;
+                    break;
+                }
+            }
+
+            if (!isFingerprintAuthenticated) {
+                throw new NullPointerException("Device fingerprint does not belong to account");
+            }
+        } finally {
+            conn.dispose();
+        }
+        return isFingerprintAuthenticated;
+    }
+
+    private static void writeAttestationHistoryJson(final HttpExchange exchange, final String deviceFingerprint,
+            final long offsetId, final int directive) throws IOException, SQLiteException {
+        final SQLiteConnection conn = new SQLiteConnection(AttestationProtocol.ATTESTATION_DATABASE);
+        final JsonArrayBuilder attestations = Json.createArrayBuilder();
+        final SQLiteStatement history;
+        try {
+            open(conn, true);
+            switch(directive) {
+                case ATTESTATION_HISTORY_PAGINATE_BACK:
+                    history = conn.prepare("SELECT time, strong, teeEnforced, " +
+                        "osEnforced, id FROM Attestations WHERE fingerprint = hex(?) AND id >= ? ORDER BY id DESC LIMIT" +
+                        String.valueOf(HISTORY_PER_PAGE));
+                    break;
+                case ATTESTATION_HISTORY_PAGINATE_NEXT:
+                    history = conn.prepare("SELECT time, strong, teeEnforced, " +
+                        "osEnforced, id FROM Attestations WHERE fingerprint = hex(?) AND id <= ? ORDER BY id DESC LIMIT" +
+                        String.valueOf(HISTORY_PER_PAGE));
+                    break;
+            }
+            history.bind(1, deviceFingerprint);
```

Comment:
Since the `switch` statement above doesn't have a default case, this fails to build because `history` can be uninitialized. You should probably just use an `if` statement above with an `else` statement handling the other possible value. It has already been verified to be either 0 or 1.